AFQ/tasks/decorators.py

```diff
@@ -0,0 +1,160 @@
+        for ii, arg_name in enumerate(list(
+                inspect.signature(func).parameters.keys())):
+            if inside:
+                if name in arg_name:
+                    vals.append(args[ii])
+                    break
+            else:
+                if arg_name == name:
+                    vals.append(args[ii])
+                    break
+
+    if len(vals) < len(names):
+        raise NameError(f"{names} requested but {vals} found in get_args")
+    return vals
+
+
+# subses_dict must be in arguments
+# if including tracking or segmentation params in fname,
+# also include them in args
+def as_file(suffix, include_track=False, include_seg=False):
+    def _as_file(func):
+        @functools.wraps(func)
+        @has_args(func)
+        def wrapper_as_file(*args, **kwargs):
+            subses_dict = get_args(
+                func, ["subses_dict"], args)[0]
+            if include_track:
+                tracking_params = get_args(
+                    func, ["tracking_params"], args)[0]
+            else:
+                tracking_params = None
+            if include_seg:
+                segmentation_params = get_args(
+                    func, ["segmentation_params"], args)[0]
+            else:
+                segmentation_params = None
+            this_file = get_fname(
+                subses_dict, suffix,
+                tracking_params=tracking_params,
+                segmentation_params=segmentation_params)
+            if not op.exists(this_file):
+                img_trk_or_csv, meta = func(*args, **kwargs)
+
+                logger.info(f"Saving {this_file}")
+                if isinstance(img_trk_or_csv, nib.Nifti1Image):
+                    nib.save(img_trk_or_csv, this_file)
+                elif isinstance(img_trk_or_csv, StatefulTractogram):
+                    save_tractogram(
+                        img_trk_or_csv, this_file, bbox_valid_check=False)
+                else:
+                    img_trk_or_csv.to_csv(this_file)
+                meta_fname = get_fname(
+                    subses_dict, suffix.split('.')[0] + '.json',
+                    tracking_params=tracking_params,
+                    segmentation_params=segmentation_params)
+                afd.write_json(meta_fname, meta)
+            return this_file
+        return wrapper_as_file
+    return _as_file
+
+
+# args must include dwi_affine
+def as_model(func):
+    @functools.wraps(func)
+    @has_args(func)
+    def wrapper_as_model(*args, **kwargs):
+        dwi_affine = get_args(
+            func, ["dwi_affine"], args)[0]
+        start_time = time()
+        model, meta = func(*args, **kwargs)
+        meta['timing'] = time() - start_time
+        model_img = nib.Nifti1Image(
+            model,
+            dwi_affine)
+        return model_img, meta
+    return wrapper_as_model
+
+
+# args must include dwi_affine
+# args must include arg with name that contains "param"
+def as_tf_deriv(tf_name):
```

Comment:
I don't like the name "as_tf_deriv", because `tf` is also what we import templateflow as and whenever I see those letters I inevitably think about tensorflow. Could we be more explicit here? "TF" always stands for "tensor fit"? Maybe "as_dt_deriv" (where "dt" = "diffusion tensor")?

Reply:
as_dt_deriv works